### PR TITLE
Initial impl of transaction parsing, make grammar more whitespace friendly

### LINKED ISSUE #5: Parse transactions (and all constructs that make up a transaction)
We've [started implementing parsing for transaction directives](https://github.com/twilco/beancount/blob/9a6c28ba6ee7deb89f0377db6e23f3292605a89a/src/beancount.pest#L101), but it is not complete.  The parse should match up to the constructs we've defined, such as [Transaction](https://github.com/twilco/beancount/blob/9a6c28ba6ee7deb89f0377db6e23f3292605a89a/src/constructs.rs#L662) and [Posting](https://github.com/twilco/beancount/blob/9a6c28ba6ee7deb89f0377db6e23f3292605a89a/src/constructs.rs#L458).  Make changes to these structures if necessary.  See the [grammar for transactions in the original Beancount for a reference](https://bitbucket.org/blais/beancount/src/487cb9b0248427ac6882889b6b8bcc1665ce393e/beancount/parser/grammar.y?at=default&fileviewer=file-view-default#grammar.y-367).

### LINKED ISSUE #10: Figure out what the '#' transaction flag does and document accordingly
As mentioned in the title, we should figure out what the `#` transaction flag does and document our code accordingly.  Currently the flag is called `flag_hash`, rather than some of the other more useful flag names such as `flag_warning` or `flag_okay`.  [It is supported in the original grammar](https://bitbucket.org/blais/beancount/src/default/beancount/parser/grammar.y?fileviewer=file-view-default#grammar.y-274), but I'm not sure what it is.  Once we discover what this flag is, let's update that grammar rule name and document the flag.

### PR DESCRIPTION
**Pull request by twilco**
This PR adds the ability to parse "happy path" transactions, in that I've only tested very basic transactions so far.  Also, this commit makes our grammar much more whitespace friendly, in that you can have whitespaces in between tokens without failing the parse.  Along this same line, we switch from Pest's built-in `WHITE_SPACE` rule to our `ws` own  rule, since I have discovered Pest's built-in `WHITE_SPACE` rule includes newline characters, which I didn't realize until now (although that behavior makes sense from Pest's perspective, moreso just my slip up).

Closes #5 and closes #10.

### CODE REVIEW

**Comment by twilco** [comment]
@mbudde - want to review?

File: src/beancount.pest [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
-eol = { NEWLINE | comment ~ NEWLINE }
-key = { !":" ~ ASCII_ALPHA_LOWER ~ (ASCII_ALPHANUMERIC | DASH | "_")* }
+indent = { ws+ }
+comment = { ws* ~ ";" ~ (!NEWLINE ~ ANY)* }
+eol = { ws* ~ NEWLINE | comment ~ NEWLINE }
+indented_comment_line = { indent ~ comment ~ NEWLINE }
+key = { ASCII_ALPHA_LOWER ~ (ASCII_ALPHANUMERIC | DASH | "_")* }
 key_value_value = { quoted_str | account | date | commodity | tag | bool | amount | num_expr | str }
-key_value = { key ~ ":" ~ WHITE_SPACE? ~ key_value_value }
+key_value = { key ~ ws* ~ ":" ~ ws* ~ key_value_value }
```
Comment (mbudde) [inline src/beancount.pest]: I don't think there can be whitespace before the colon?

File: src/beancount.pest [hunk 2]
```diff
@@ -42,61 +45,116 @@ commodity = @{ ASCII_ALPHA_UPPER ~ commodity_trailing{1, 23} }
...
 
 // 2012-01-01 commodity HOOL
-commodity_directive = { date ~ WHITE_SPACE ~ "commodity" ~ WHITE_SPACE ~ commodity ~ eol ~ key_value_list? }
+commodity_directive = { date ~ ws+ ~ "commodity" ~ ws+ ~ commodity ~ eol ~ key_value_list? }
 
 // 2014-07-09 custom "budget" "some_config_opt_for_custom_directive" TRUE 45.30 USD
 custom_value = { quoted_str | date | bool | amount | num_expr | account }
-custom_value_list = { custom_value ~ (WHITE_SPACE ~ custom_value)* }
-custom = { date ~ WHITE_SPACE ~ "custom" ~ WHITE_SPACE ~ quoted_str ~ WHITE_SPACE ~ custom_value_list? ~ eol ~ key_value_list? }
+custom_value_list = { custom_value ~ (ws ~ custom_value)* }
```
Comment (mbudde) [inline src/beancount.pest]: Should that be `ws+`?

File: src/beancount.pest [hunk 3]
```diff
@@ -42,61 +45,116 @@ commodity = @{ ASCII_ALPHA_UPPER ~ commodity_trailing{1, 23} }
...
+    indent ~ (txn_flag ~ ws+)? ~ account ~ eol
+}
+posting_or_kv_list = {
+  (key_value_line | posting | (indent ~ tags_links ~ NEWLINE) | (indented_comment_line))*
+}
+price_annotation = { incomplete_amount }
+// 189.10 USD | USD | 152.41
+incomplete_amount = { (num_expr ~ ws+ ~ commodity) | commodity | num_expr }
+// {{502.12 # 9.95 USD, 2018-01-01}}
+cost_spec = { ("{{" | "{") ~ cost_comp_list ~ ("}}" | "}") }
```
Comment (mbudde) [inline src/beancount.pest]: This matches unbalanced braces: `{{ ... }`. Should either be split into two rules or use `PUSH` and `POP`.

Comment (twilco) [reply src/beancount.pest]: Ahh, great catch.  Fixed.

File: src/beancount.pest (hunk 2 above)
Comment (twilco) [reply src/beancount.pest]: Yes!

File: src/beancount.pest (hunk 1 above)
Comment (twilco) [reply src/beancount.pest]: Correct - fixing.

**Comment by twilco** [comment]
I did indeed try using Pest's implicit whitespace rule, but things got funky with newlines.  I didn't really give it a fair shake though, and it would probably be more robust than what we're doing currently, so I think I'll give it another go in a separate PR.

**Comment by twilco** [comment]
@mbudde 

I looked into using Pest's implicit whitespace.  I'm not sure we would benefit much since the original grammar is whitespace-sensitive with indents required in key-value lists, postings, etc., which would, in turn, pervade to (I think) a majority of our rules.  All rules using any of these whitespace-sensitive would need to be atomic (the `@` or `$` symbol).  Making a rule atomic eliminates all implicit whitespace, meaning we would have to manually specify all our whitespace just as we are doing now.

This is my first foray into Pest, so maybe I'm missing something here.  The original grammar appears to be implicitly whitespaced, so it must be possible somehow 🙂 .  Perhaps I'll give it another go later.

You do raise a great point in that we currently do not support syntax that has no spaces even though it is valid (`2019-02-19*"Foo""Bar"`), so I'll create an issue for that.

**Pull request merged**